http://bugs.gentoo.org/144887
http://gcc.gnu.org/PR28634

2006-09-09  Richard Sandiford  <richard@codesourcery.com>

	PR rtl-optimization/28634
	* reorg.c (fill_slots_from_thread): Do not assume A + X - X == A
	for floating-point modes unless flag_unsafe_math_optimizations.

--- branches/gcc-4_1-branch/gcc/reorg.c	2006/09/09 00:27:47	116795
+++ branches/gcc-4_1-branch/gcc/reorg.c	2006/09/09 10:56:31	116796
@@ -2867,6 +2867,8 @@
       dest = SET_DEST (pat), src = SET_SRC (pat);
       if ((GET_CODE (src) == PLUS || GET_CODE (src) == MINUS)
 	  && rtx_equal_p (XEXP (src, 0), dest)
+	  && (!FLOAT_MODE_P (GET_MODE (src))
+	      || flag_unsafe_math_optimizations)
 	  && ! reg_overlap_mentioned_p (dest, XEXP (src, 1))
 	  && ! side_effects_p (pat))
 	{
